### PR TITLE
Use DMA Engine API for DMA transactions

### PR DESCRIPTION
**Pull request by jcdevel**
I had to create this patch in order to build the Logibone kernel module against Kernel 4.1 from:
https://github.com/beagleboard/linux/tree/4.1
Feedback is welcome.
Thanks,
JC

### CODE REVIEW

**Comment by jcdevel** [comment]
I have just realized that I have updated only the Makefile for Logibone R1, since that is the board I am using. My patch will break the build for other boards, but the fix is simple (just including the new logi_dma.c source file in the corresponding Makefile).

**Comment by QuickJack** [comment]
The existing code consists of main_dm.c and main_dma.c. Jonathan had problems with the DMA part. Because of that the official Logibone (R1) module doesn't use DMA. The makefile for Logibone DMA generates two .ko versions for DMA/ non DMA variants, respectively. I am using the DMA variant ever since without any problems.

What you did is basically a mix of the already existing code. You are using #define to distinguish between DMA/non DMA variants instead of different modules.

What I do not understand is, why you are trying to use a different kernel. The bb-kernel is working perfectly even when using the PREEMPT_RT variant. I can say that the performance has been improved greatly.

**Comment by QuickJack** [comment]
By looking at your modifications I can see that you moved the DMA related stuff to a new module called logi_dma.c. Based on LOGI_USE_DMAENGINE you are switching between the already existing code and a new variant.

Robert told me something about an EDMA rewrite. Is this the reason for the new variant?

**Comment by jcdevel** [comment]
Being honest, I started rewriting the module just for learning purposes (you could call it stubbornness), but then I thought that if TI has removed the EDMA API from the Kernel BSP, at some point the bb-kernel will have to catch up with that (or maybe not, I do not really know what is the policy here). Regarding PREEMPT_RT, if I ever need real time tasks, my plan is using the PRU available in the AM3358 SoC, leaving the Cortex-A for other tasks, so I do not really need PREEMPT_RT support.

I noticed there are two versions of the module: one in the 'beaglebone-black' directory, the other inside 'KERNEL'. I do not know which one is the official, I do not even know why there are two of them, I just picked one as the base for my patch.

I think that with a bit of effort all modules could be merged together in one single location, improving code reusability. Features could be selected at build time (DMA, no DMA, EDMA API, DMA Engine API, etc), This patch is just a first attempt towards that direction.

**Comment by jcdevel** [comment]
BTW, LOGI_USE_DMAENGINE does not switch between DMA/non DMA variants, it switches between EDMA API/DMA Engine API. The module is the same for both APIs (it is a build option).

File: beaglebone-black/common/main_dma.c [hunk 1]
```diff
@@ -242,22 +234,10 @@ static int dm_open(struct inode *inode, struct file *filp)
...
-			DBG_LOG("EDMA channel %d reserved\n", mem_dev->dma_chan);
-			mem_dev->dma_buf = (unsigned char *) dma_alloc_coherent(NULL, MAX_DMA_TRANSFER_IN_BYTES, &dmaphysbuf, 0);
 
-			if (mem_dev->dma_buf == NULL) {
-				DBG_LOG("failed to allocate DMA buffer\n");
-
-				return -ENOMEM;
-			}
+			result = logi_dma_init(mem_dev, &dmaphysbuf);
+			if (result)
```
Comment (QuickJack) [inline beaglebone-black/common/main_dma.c]: result is negative in case of an error! => should be changed to "if (result < 0)"

**Comment by QuickJack** [comment]
I have to double check that the existing code isn't changed by your patch. The new code is activated by defining LOGI_USE_DMAENGINE. There is a problem when checking the return value for logi_dma_init(). If all that is satisfied then I think your code is ready to be merged. I will adjust all other sources based on your changes afterwards.

The original code is in https://github.com/fpga-logi/logi-kernel/tree/master/beaglebone-black. These modules are using the original FPGA configuration method by using dd on /dev/logibone. The code is built as separate kernel modules which allows for maximum flexibility. As a developer I do want to have just that. 

But not all people are familiar with Linux kernel development. The KERNEL subfolder contains code that is to be merged directly into the bb-kernel. The idea is, that the Logibone would be supported out of the box by the official bb-kernel. However, it seems impossible to keep the changes up to date. Also these changes have been merged exclusively into the 3.8.13 branch. Besides of that, they also changed the FPGA configuration method. The code in this subfolder requires a separate configuration utility but shares the same main_dma.c module. I never liked that concept.

**Comment by jcdevel** [comment]
Maybe I am talking nonsense, but could we add an entry in the Device Tree specific to the Logibone kernel module? For example, we could specify the DMA configuration or the FPGA configuration method (even link it to the pinout configuration). Then the source code could be contained in a single kernel module, easier to understand and maintain. Maybe we can hold this patch until we make a decision? Regarding the changes, I have relocated a couple of fields in the structs in drvr.h (the DMA buffer and channel number). Other than that, I think the existing EDMA code is mostly the same.

The thing is, I do not like the idea behind the bb-kernel scripts. Reapplying patches makes very difficult to track the branch history, or figure out which features are present and which aren't using basic git commands (please understand that I am not an advanced git user). I mean, if TI (or any developer) already provides a branch with a clean and traceable log history, I do not see the need of picking up individual patches and reapply them on a Linux stable branch, creating two similar but disconnected branches. Let's say I want to know what patches from TI have not been applied to the bb-kernel. Having two disconnected branches, it is nearly impossible to obtain for the regular git user. This is one of the reasons I am using the BeagleBone kernel, although I it would be nice if the Beagle specific patches were merged instead of constantly rebase them every time there is a new release.

**Comment by QuickJack** [comment]
I have rebased my repo against yours and implemented all required changes to support all variants without breaking existing functionality. You are welcome to help testing. Jonathan is not responding at the moment so we have to keep our repos in sync. The pull requests will be merged later into fpga-logi.

Device trees have many capabilities. But unfortunately, they are also very difficult to understand. In the logi-kernel project they constantly caused a lot of trouble. The underlying concepts are changing with each kernel release. I am not sure if they could help on improving the readability and understanding of the source code.

The code is split into several modules to allow for a better coding structure. In many applications it is necessary to create a custom driver. This can easily be accomplished by modifying the ioctl module only.

**Pull request merged**